### PR TITLE
StringReverseDataset character version and variable length support

### PR DESCRIPTION
**Pull request by TahaZahid05**
## Overview
This PR updates the `StringReverseDataset` to use characters instead of integers and adds support for variable-length sequences.

### Dataset & Model
- **Alphabet-based data**: The dataset now generates sequences using a configurable alphabet (default `a-z`).
- **Variable Lengths**: Added `min_seq_len`, `max_seq_len`, and `fixed_length` settings.
- **Padding**: Reserved Index `0` for padding. Added masking in `CrossEntropyLoss` and Accuracy so padding tokens are ignored during training and evaluation.
- **Decoding**: Added a `.decode()` method to convert IDs back into characters.

### Framework & CLI
- **Modular Config**: Updated the `PreTrainer` to take `config.dataset` directly.
- **Pydantic Validation**: Updated type-hints in `ProjectConfig` so YAML configs load correctly into objects.
- **CLI Initialization**: Updated the `trainite init` logic to handle the new function calls when scaffolding projects.
- **Example Update**: Updated the `examples/string-reverse` project to use these changes.

### CODE REVIEW

**Comment by TahaZahid05** [comment]
Currently, `base.py` uses specific type-hints:

https://github.com/pytorch-ignite/trainite/blob/8e27f1f583b4386fcf7d931d93f0d8161bced814/trainite/config/base.py#L21-L23

We need a more generic validation strategy later on so it remains model agnostic.

File: examples/string-reverse/config.py [hunk 1]
```diff
@@ -15,12 +15,15 @@ class TransformerModelConfig(BaseModel):
     max_seq_len: int = 128
 
 class StringReverseDatasetConfig(BaseModel):
-    vocab_size: int = 32
+    alphabet: str = "abcdefghijklmnopqrstuvwxyz"
```
Comment (aaishwarymishra) [inline examples/string-reverse/config.py]: I am not sure we should take the alphabets or not, we can just take type, whether alpha or alpha numeric.

File: examples/string-reverse/config.py [hunk 2]
```diff
@@ -15,12 +15,15 @@ class TransformerModelConfig(BaseModel):
...
+    vocab_size: int = 26
```
Comment (aaishwarymishra) [inline examples/string-reverse/config.py]: Also whats the use of vocab_size here? Its understandable in the model but not sure in the dataset.

File: examples/string-reverse/dataset.py [hunk 3]
```diff
@@ -2,56 +2,115 @@
...
+    def decode(self, ids: torch.Tensor) -> str:
+        return "".join([self.id_to_char[idx.item()] for idx in ids if idx.item() != 0])
+
+
+def collate_fn(batch: list[dict[str, torch.Tensor]]) -> dict[str, torch.Tensor]:
+    input_ids = [item["input_ids"] for item in batch]
+    labels = [item["labels"] for item in batch]
+    
+    padded_input_ids = pad_sequence(input_ids, batch_first=True, padding_value=0)
+    padded_labels = pad_sequence(labels, batch_first=True, padding_value=0)
```
Comment (aaishwarymishra) [inline examples/string-reverse/dataset.py]: I think we should take default padding of -100 , instead of the 0, it would also made us stop using vocab_size+1

File: examples/string-reverse/config.py (hunk 1 above)
Comment (TahaZahid05) [reply examples/string-reverse/config.py]: Good point. Maybe we should give both options? The strings can provide flexibility incase they want to test against a specific subset of characters but maybe utitlies like: `alphabet: "@digits"` or `alphabet: "@lowercase"` for shortcuts to help with the issue you have pointed out. What do you think?

File: examples/string-reverse/dataset.py (hunk 3 above)
Comment (TahaZahid05) [reply examples/string-reverse/dataset.py]: True, default padding should be -100. I'll fix that. But curious, how would this lead to stop using vocab_size + 1? We would still need that becaues `nn.Embedding` require a positive index for padding meaning we need the 0th index for it.

File: examples/string-reverse/config.py (hunk 1 above)
Comment (aaishwarymishra) [reply examples/string-reverse/config.py]: I am not sure I understand what you said.

File: examples/string-reverse/dataset.py (hunk 3 above)
Comment (aaishwarymishra) [reply examples/string-reverse/dataset.py]: when we are using 0 as padding, 0 will not be considered an index of a character , with 0 as padding vocab index stars from 1, and the vocab size would be vocab+1

File: examples/string-reverse/config.py (hunk 1 above)
Comment (TahaZahid05) [reply examples/string-reverse/config.py]: I am saying user should have option of both, adding custom alphabets and some presets. like:

```python
# Standard presets
alphabet: "@alpha"    # a-z
alphabet: "@digits"   # 0-9

# Or custom research sets
alphabet: "abc!@#"   
```

File: examples/string-reverse/dataset.py (hunk 3 above)
Comment (TahaZahid05) [reply examples/string-reverse/dataset.py]: Can you show through example? I think it doesn't matter whether you use -100 or 0 as label, the input still needs a number to act as a placeholder for the padding in a batch.

Since that placeholder number can't be one of our 26 letters, we always end up needing 27 slots (26 letters + 1 placeholder).

Comment (aaishwarymishra) [reply examples/string-reverse/dataset.py]: here https://github.com/pytorch-ignite/trainite/pull/3/changes#diff-e0dac8bceead0c2b475a6440a83d3cb3f966218558961b9832d2eeb75dbb4250R21-R23

So basically you were using the 0 as padding your starting index for mapping is 0+1, if you had say 6 strings the last index will be 5+1, as last index is 6 so your vocab_size is 6.
Embedding layer expects the indexes or ids to start from 0, so if your vocab_size is 6 it expects valid indexes to be in between 0-5, here we have reserved 0 for padding so valid index for us will be 1-6, thus we need to set vocab_size+1.

Comment (TahaZahid05) [reply examples/string-reverse/dataset.py]: I understand that.. How does using `-100` solve this? You still require an index for padding which results in vocab_size+1?

Comment (aaishwarymishra) [reply examples/string-reverse/dataset.py]: Pytorch layers ignore -100 by default as they treat it as padding . So you do not  need to explicitly do vocab_size + 1.

Comment (TahaZahid05) [reply examples/string-reverse/dataset.py]: `nn.Embedding` doesn't ignore it, which means we still have to give it a space. `nn.CrossEntropyLoss` ignores it thats why I changed to -100. I think you are confusing Loss and Embedding.

Comment (aaishwarymishra) [reply examples/string-reverse/dataset.py]: Yeah my bad.

File: examples/string-reverse/config.py (hunk 1 above)
Comment (aaishwarymishra) [reply examples/string-reverse/config.py]: sounds good

File: examples/string-reverse/dataset.py (hunk 3 above)
Comment (aaishwarymishra) [inline examples/string-reverse/dataset.py]: For labels the padding value should be -100, if taking 0 we have to set it explicitly in loss function in the trainer

File: trainite/models/transformer.py [hunk 4]
```diff
@@ -69,7 +69,7 @@ def __init__(
         **kwargs,
     ) -> None:
         super().__init__()
-        self.embedding = nn.Embedding(vocab_size, hidden_size)
+        self.embedding = nn.Embedding(vocab_size + 1, hidden_size, padding_idx=0)
```
Comment (aaishwarymishra) [inline trainite/models/transformer.py:72]: As default, should we give user the option to change padding or enable or disable by default?  I am not sure

File: trainite/datasets/string_reverse.py [hunk 5]
```diff
@@ -1,60 +1,126 @@
...
+        self,
+        size: int,
+        min_seq_len: int,
+        max_seq_len: int,
+        seed: int,
+        fixed_length: bool = True,
+        alphabet: str = "abcdefghijklmnopqrstuvwxyz",
+    ) -> None:
+        self.alphabet = ALPHABET_PRESETS.get(alphabet, alphabet)
+        vocab_size = len(self.alphabet)
```
Comment (aaishwarymishra) [inline trainite/datasets/string_reverse.py:24]: Here vocab_size in dynamically set but in model we have to set it manually, I am not sure if its correct pattern.

**Comment by aaishwarymishra** [comment]
I just noticed at root there is `config.py`, its not getting used anywhere I may have forgot to remove it.

File: trainite/datasets/string_reverse.py (hunk 5 above)
Comment (TahaZahid05) [reply trainite/datasets/string_reverse.py:24]: Well, that is because dataset's vocab_size is derived from the `alphabet` string. Whereas, model's vocab_size is the capacity of the model's embedding as to the maximum string length it can process. It is not something we can add dynamically based on the dataset as models can be trained on certain lengths of strings and then tested on even larger strings to check for OOD. Does that make sense?

Comment (aaishwarymishra) [reply trainite/datasets/string_reverse.py:24]: what about say we define vocab size in datasets, say 64. During the string generation phase it takes first 64 characters from the alphabet like `alphabet[:64]`. I am not sure if its a good idea or not though.

File: trainite/models/transformer.py (hunk 4 above)
Comment (TahaZahid05) [reply trainite/models/transformer.py:72]: I don't think we should give option to disable padding as that is a critical requirement for batching in model training.. There is no use of disabling padidngs?

Also, for changing padding. I don't see how that can affect training or be useful in experiments. Your thoughts?

Comment (aaishwarymishra) [reply trainite/models/transformer.py:72]: I think user can add it himself if he wants so its fine ig

File: trainite/datasets/string_reverse.py (hunk 5 above)
Comment (TahaZahid05) [reply trainite/datasets/string_reverse.py:24]: That can be quite ambiguous because what if the user wants certain parts of the `alphabet` in their `vocab_size` defined rather than just the first 64. I think the most flexible approach is that they use the shorthands or specify the characters they want to be used and then we calculate length dynamically based on their input.

Comment (aaishwarymishra) [reply trainite/datasets/string_reverse.py:24]: My main concern is for the model, user wont know exact vocab size needed for training the model if its getting calculated dynamically, either he has to set the value high, or dynamically get the vocb_size

Comment (TahaZahid05) [reply trainite/datasets/string_reverse.py:24]: Good point. Maybe, we can write `vocab_size` for dataset into the yaml but explicitly mark it as a derived value by adding a comment, and also some validation to see if it matches the desired length. This  way the user can see the dataset's `vocab_size` so they can set model's `vocab_size` accordingly. 

Thoughts?

**Comment by TahaZahid05** [comment]
@aaishwarymishra as a temporary approach, I have made the model's `vocab_size` default value point to dataset's `vocab_size`. This approach still has cons of the user manually changing it if they decide to modify `alphabet`. We may need to discuss a better approach together in tomorrow's meeting. I would suggest we can merge this for now and deal with the design issue later after we talk about it.

**Comment by aaishwarymishra** [comment]
Okay you can merge it whenever it's ready 👍
